File: consorcio/lib/domain/usecases/list_activity_published_work_use_case.dart
import '../entities/app_user.dart';
import '../entities/folder_date.dart';
import '../entities/technician_activity_work.dart';
import '../errors/domain_exception.dart';
import '../repositories/area_repository.dart';
import '../repositories/folder_date_repository.dart';
import '../repositories/folder_image_repository.dart';
import '../repositories/image_folder_repository.dart';
import '../repositories/user_repository.dart';
import '../value_objects/user_role.dart';

class ListActivityPublishedWorkUseCase {
  ListActivityPublishedWorkUseCase(
    this._areaRepository,
    this._folderRepository,
    this._dateRepository,
    this._imageRepository,
    this._userRepository,
  );

  final AreaRepository _areaRepository;
  final ImageFolderRepository _folderRepository;
  final FolderDateRepository _dateRepository;
  final FolderImageRepository _imageRepository;
  final UserRepository _userRepository;

  Future<ActivityPublishedWorkResult> execute(
    AppUser actor,
    String areaId,
  ) async {
    actor.assertCanOperateApp();
    final trimmed = areaId.trim();
    if (trimmed.isEmpty) {
      throw DomainException('Actividad inválida');
    }

    final area = await _areaRepository.getById(trimmed);
    if (area == null) {
      throw DomainException('Actividad no encontrada');
    }

    final folders = actor.isMobileAdmin
        ? await _folderRepository.listByArea(trimmed)
        : await _folderRepository.listAccessibleForUserAndArea(
            userId: actor.id,
            areaId: trimmed,
          );
    final technicians = await _userRepository.listTechnicians();
    final folderIds = folders.map((folder) => folder.id).toList();

    final allImages = await _imageRepository.listByFolderIds(folderIds);
    final images = actor.isMobileAdmin
        ? allImages
        : allImages
            .where((image) => image.uploadedById == actor.id)
            .toList();
    final dates = await _dateRepository.listByFolderIds(folderIds);
    final folderById = {for (final folder in folders) folder.id: folder};
    final dateById = {for (final date in dates) date.id: date};

    final worksByKey = <String, PublishedTechnicianWork>{};
    for (final image in images) {
      final technicianId = image.uploadedById.trim();
      if (technicianId.isEmpty) continue;
      final dateId = image.dateId.trim();
      if (dateId.isEmpty) continue;
      final folder = folderById[image.folderId];
      if (folder == null) continue;
      final dateKey = dateById[image.dateId]?.dateKey ??
          FolderDate.toDateKey(image.createdAt);
      final key = '$technicianId|${image.folderId}|$dateId';
      final existing = worksByKey[key];
      if (existing != null) {
        final later = image.createdAt.isAfter(existing.publishedAt);
        worksByKey[key] = PublishedTechnicianWork(
          technicianId: existing.technicianId,
          technicianName: existing.technicianName,
          folderId: existing.folderId,
          dateId: existing.dateId,
          routeCode: existing.routeCode,
          folderName: existing.folderName,
          dateKey: existing.dateKey,
          imageCount: existing.imageCount + 1,
          publishedAt: later ? image.createdAt : existing.publishedAt,
        );
        continue;
      }
      worksByKey[key] = PublishedTechnicianWork(
        technicianId: technicianId,
        technicianName: image.uploadedByName.trim().isEmpty
            ? 'Técnico'
            : image.uploadedByName.trim(),
        folderId: image.folderId,
        dateId: dateId,
        routeCode: folder.routeCode ?? '',
        folderName: folder.name,
        dateKey: dateKey,
        imageCount: 1,
        publishedAt: image.createdAt,
      );
    }

    final works = worksByKey.values.toList()
      ..sort((left, right) {
        final byDate = right.dateKey.compareTo(left.dateKey);
        if (byDate != 0) return byDate;
        return (left.routeCode.isEmpty ? left.folderName : left.routeCode)
            .compareTo(
          right.routeCode.isEmpty ? right.folderName : right.routeCode,
        );
      });

    final stats = <String, ActivityTechnicianFolder>{};
    for (final work in works) {
      final current = stats[work.technicianId];
      final last = current?.lastPublishedAt;
      stats[work.technicianId] = ActivityTechnicianFolder(
        technicianId: work.technicianId,
        technicianName: work.technicianName,
        workCount: (current?.workCount ?? 0) + 1,
        imageCount: (current?.imageCount ?? 0) + work.imageCount,
        lastPublishedAt: last == null || work.publishedAt.isAfter(last)
            ? work.publishedAt
            : last,
      );
    }

    final technicianFolders = <String, ActivityTechnicianFolder>{};
    for (final technician in technicians) {
      if (!technician.active) continue;
      if (!technician.assignedRoles.contains(UserRole.tecnico)) continue;
      final current = stats[technician.id];
      technicianFolders[technician.id] = ActivityTechnicianFolder(
        technicianId: technician.id,
        technicianName: technician.displayName,
        workCount: current?.workCount ?? 0,
        imageCount: current?.imageCount ?? 0,
        lastPublishedAt: current?.lastPublishedAt,
      );
    }
    for (final entry in stats.entries) {
      technicianFolders.putIfAbsent(entry.key, () => entry.value);
    }

    final technicianList = technicianFolders.values.toList()
      ..sort((left, right) {
        if (right.workCount != left.workCount) {
          return right.workCount.compareTo(left.workCount);
        }
        return left.technicianName.compareTo(right.technicianName);
      });

    return ActivityPublishedWorkResult(
      areaId: area.id,
      areaName: area.name,
      technicians: technicianList,
      works: works,
    );
  }
}

File: src/domain/usecases/folders/ListActivityPublishedWorkUseCase.ts
import type { User } from '@/domain/entities/User'
import { assertUserCanAccessFolder } from '@/domain/entities/User'
import type { AreaRepository } from '@/domain/repositories/AreaRepository'
import type { FolderDateRepository } from '@/domain/repositories/FolderDateRepository'
import type { FolderImageRepository } from '@/domain/repositories/FolderImageRepository'
import type { ImageFolderRepository } from '@/domain/repositories/ImageFolderRepository'
import type { UserRepository } from '@/domain/repositories/UserRepository'
import { toDateKey } from '@/domain/entities/FolderDate'
import type {
  ActivityPublishedWorkResult,
  ActivityTechnicianFolder,
  PublishedTechnicianWork,
} from '@/domain/entities/TechnicianActivityWork'
import { UnauthorizedError, ValidationError } from '@/domain/errors/DomainError'
import { hasAssignedRole, UserRole } from '@/domain/value-objects/UserRole'

function workKey(technicianId: string, folderId: string, dateId: string): string {
  return `${technicianId}|${folderId}|${dateId}`
}

export class ListActivityPublishedWorkUseCase {
  private readonly areaRepository: AreaRepository
  private readonly folderRepository: ImageFolderRepository
  private readonly dateRepository: FolderDateRepository
  private readonly imageRepository: FolderImageRepository
  private readonly userRepository: UserRepository

  constructor(
    areaRepository: AreaRepository,
    folderRepository: ImageFolderRepository,
    dateRepository: FolderDateRepository,
    imageRepository: FolderImageRepository,
    userRepository: UserRepository,
  ) {
    this.areaRepository = areaRepository
    this.folderRepository = folderRepository
    this.dateRepository = dateRepository
    this.imageRepository = imageRepository
    this.userRepository = userRepository
  }

  async execute(
    actor: User,
    areaId: string,
  ): Promise<ActivityPublishedWorkResult> {
    if (!actor.active) {
      throw new UnauthorizedError('Cuenta inactiva')
    }
    const trimmedAreaId = areaId.trim()
    if (!trimmedAreaId) {
      throw new ValidationError('Actividad inválida')
    }

    const area = await this.areaRepository.getById(trimmedAreaId)
    if (!area) {
      throw new ValidationError('Actividad no encontrada')
    }

    const [folders, technicians] = await Promise.all([
      this.folderRepository.listByArea(trimmedAreaId),
      this.userRepository.listTechnicians(),
    ])

    const visibleFolders = folders.filter((folder) =>
      assertUserCanAccessFolder(actor, folder),
    )
    const folderIds = visibleFolders.map((folder) => folder.id)

    const [images, dates] = await Promise.all([
      this.imageRepository.listByFolderIds(folderIds),
      this.dateRepository.listByFolderIds(folderIds),
    ])

    const folderById = new Map(visibleFolders.map((folder) => [folder.id, folder]))
    const dateById = new Map(dates.map((item) => [item.id, item]))

    const worksByKey = new Map<
      string,
      PublishedTechnicianWork & { latest: Date }
    >()

    for (const image of images) {
      const technicianId = image.uploadedById.trim()
      if (!technicianId) continue
      const dateId = image.dateId.trim()
      if (!dateId) continue
      const folder = folderById.get(image.folderId)
      if (!folder) continue
      const folderDate = dateById.get(image.dateId)
      const dateKey =
        folderDate?.dateKey || toDateKey(image.createdAt)
      const key = workKey(technicianId, image.folderId, dateId)
      const existing = worksByKey.get(key)
      if (existing) {
        existing.imageCount += 1
        if (image.createdAt.getTime() > existing.latest.getTime()) {
          existing.latest = image.createdAt
          existing.publishedAt = image.createdAt
        }
        continue
      }

      worksByKey.set(key, {
        technicianId,
        technicianName: image.uploadedByName.trim() || 'Técnico',
        folderId: image.folderId,
        dateId,
        routeCode: folder.routeCode || '',
        folderName: folder.name,
        dateKey,
        imageCount: 1,
        publishedAt: image.createdAt,
        latest: image.createdAt,
      })
    }

    const works = [...worksByKey.values()]
      .map(({ latest: _latest, ...work }) => work)
      .sort((left, right) => {
        const byDate = right.dateKey.localeCompare(left.dateKey)
        if (byDate !== 0) return byDate
        return (left.routeCode || left.folderName).localeCompare(
          right.routeCode || right.folderName,
        )
      })

    const statsByTechnician = new Map<
      string,
      { workCount: number; imageCount: number; lastPublishedAt: Date | null; name: string }
    >()
    for (const work of works) {
      const current = statsByTechnician.get(work.technicianId) ?? {
        workCount: 0,
        imageCount: 0,
        lastPublishedAt: null,
        name: work.technicianName,
      }
      current.workCount += 1
      current.imageCount += work.imageCount
      if (
        !current.lastPublishedAt ||
        work.publishedAt.getTime() > current.lastPublishedAt.getTime()
      ) {
        current.lastPublishedAt = work.publishedAt
      }
      if (work.technicianName) current.name = work.technicianName
      statsByTechnician.set(work.technicianId, current)
    }

    const technicianFolders = new Map<string, ActivityTechnicianFolder>()
    for (const technician of technicians) {
      if (!technician.active || !hasAssignedRole(technician, UserRole.Tecnico)) {
        continue
      }
      const stats = statsByTechnician.get(technician.id)
      technicianFolders.set(technician.id, {
        technicianId: technician.id,
        technicianName: technician.displayName,
        workCount: stats?.workCount ?? 0,
        imageCount: stats?.imageCount ?? 0,
        lastPublishedAt: stats?.lastPublishedAt ?? null,
      })
    }
    for (const [technicianId, stats] of statsByTechnician) {
      if (technicianFolders.has(technicianId)) continue
      technicianFolders.set(technicianId, {
        technicianId,
        technicianName: stats.name,
        workCount: stats.workCount,
        imageCount: stats.imageCount,
        lastPublishedAt: stats.lastPublishedAt,
      })
    }

    const technicianList = [...technicianFolders.values()].sort((left, right) => {
      if (right.workCount !== left.workCount) {
        return right.workCount - left.workCount
      }
      return left.technicianName.localeCompare(right.technicianName, 'es')
    })

    return {
      areaId: area.id,
      areaName: area.name,
      technicians: technicianList,
      works,
    }
  }
}
